546 improve documentation (#189)
adorsys/keycloak-ssi-deployment: Readme.md

Reviewer: Could you please format this file?

Author: Formated, merci !

# Keycloak as an OID4VC Issuer

Bring OpenID for Verifiable Credential Issuance (OID4VCI) to a vanilla Keycloak with a single CLI. This repo contains a tiny-but-complete toolkit: Docker compose files, scripted setup, and curl-based request flows so you can evaluate VC issuance locally.

## What You Get
- Opinionated Keycloak instance (Postgres + HTTPS) with the `oid4vc-vci` feature enabled.
- `keycloak-ssi` CLI that wraps every script under `src/` and handles configuration, setup, and testing.
- Ready-made credential definitions (IdentityCredential, SteuerberaterCredential, KMACredential) and a demo user for both pre-authorized and authorization-code flows.

## TL;DR

> Prereqs: Java 17+ (Java 21 for `import`), `openssl`, `keytool`, `jq`, `yq`, `docker`, and optionally `figlet` for pretty banners.
> - `docs/CLI_GUIDE.md` - 🚨 extended CLI reference and command list. Please take a look since it will help you understand how to interact with this repo. 

### Quick Local Spin-Up (Docker Compose)

> > - `docs/CLI_GUIDE.md` - This repo contains a CLI tool that will help you configure keycloak and run tests. Please make sure you have it before proceeding:

Reviewer: I suggest moving the optional CLI installation instructions before this section and making it clear that users can either run `./keycloak-ssi.sh` directly or install the CLI tool to make it available globally.

Don’t forget to mention that to explore the supported commands, they can use `./keycloak-ssi.sh help` or `keycloak-ssi help` if the CLI is installed.

Author: Added this section, before '### Quick Local Spin-Up (Docker Compose)'

### Optional: Install CLI Globally

A CLI tool to interact with this project was built in the keycloak-ssi.sh file. This can be installed to be available globally or you can directly interact with the scipt. To install:

```bash
./keycloak-ssi.sh install
```
To know more about the tool:

```bash
keycloak-ssi help
```

> 

Reviewer: Consider adding the following note before the commands below:
**Note:** We assume the CLI tool has been successfully installed before running these commands.

```bash
keycloak-ssi compose up -d   # boots Postgres + Keycloak with HTTPS on localhost
keycloak-ssi config          # configures realm, clients, demo users, keys
keycloak-ssi test preauth IdentityCredential
```
- Uses `config.yaml` plus overrides to drive runtime settings.

> - Admin console available at `${KEYCLOAK_ADMIN_ADDR}` (defaults to `https://localhost:8443`).

Reviewer: ```suggestion
- Admin console available at `${KEYCLOAK_ENDPOINTS_ADMIN_ADDR}` (defaults to `https://localhost:8443`).
```

Author: Improved.

### Stop / Clean
```bash

> keycloak-ssi compose down    # stop containers

Reviewer: ```suggestion
keycloak-ssi compose down -v   # stop containers
```

Author: Added.

keycloak-ssi stop            # stop a tarball-based run
```

### Optional: Install CLI Globally
```bash
./keycloak-ssi.sh install
keycloak-ssi help
```
Installs into `~/.local/bin` and symlinks project files under `~/.local/share/keycloak-ssi-deployment`.

## How It Works

```
+-------------------+     +------------------------+
| keycloak-ssi CLI  |---->| src/deployment/*.sh    |--> Keycloak (oid4vc)
| - setup/config    |     | src/credentials/*.sh   |
| - compose bridge  |     |                        |
+-------------------+     +------------------------+
         ^                        ^
         | config.yaml            | overrides/env vars

>          +------------------------+

Reviewer: I’m not entirely sure what this diagram is depicting. We could improve it by redrawing it, maybe using draw.io, and adding the updated image here. WDYT?

Author: Removed.

```

1. `setup` builds Keycloak from either an official tarball or a source branch (controlled by `keycloak.version` and `keycloak.target_branch`).
2. `compose` runs the bundled `docker-compose.yml` (Postgres + Keycloak) when you prefer containers.
3. `config` applies the realm via Keycloak Config CLI (or kcadm) scripts, creating credential definitions as client scopes.
4. `test` scripts issue credentials through curl, covering pre-authorized and authorization-code + PKCE flows.

> ## Configuration Model

Reviewer: I suggest we move this section above the `Quick Local Spin-Up (Docker Compose)` section, as we need to ensure the keystore path is set correctly based on the environment (Docker or host). It would also be good to mention the keystore section as an important part to check before running any commands.

Author: I think you properly added a comment to the config.yaml file addressing this. 

<img width="729" height="156" alt="image" src="https://github.com/user-attachments/assets/33e85719-6f7c-4c58-a785-d1eeee825ec6" />

if this is what you are referring to, I think the comment is on the right place and there is no need to put it again here.

| Layer | Description |
| --- | --- |
| `config.yaml` | Base defaults for Keycloak, database, credentials, demo users. |
| `config.override.yaml` | Optional environment-specific, overrides config.yaml. |
| Environment vars | Highest priority. Use `SECTION_KEY=value` (e.g., `KEYCLOAK_HTTPS_PORT=9443`). Values inside YAML may reference `${ENV_VAR}`. |

Key settings to review:

> Our config contains many settings to configure keycloak. The more relevant are:

Reviewer 2: ```suggestion
Our config contains many settings to configure Keycloak. The more relevant are:
```

Author: Added capital K.

- **Certificates**: `generate-kc-certs.sh` produces the HTTPS cert/key and truststore referenced in `config.yaml`. Update passwords/paths there if you reuse existing certs.
- **Client scopes = credentials**: Each credential type is modeled as a client scope (see `src/config/client-scope-config.json`). Assign the scope to clients and add `"oid4vci.enabled": "true"` in client attributes.
- **Logs & artifacts**: `target/` holds Keycloak logs, downloaded tarballs, and generated keys. **Clean it if you need a fresh start.**

## Troubleshooting
- **CLI cannot find the project**: run commands from the repo root or reinstall via `./keycloak-ssi.sh install`.
- **`keycloak-ssi import` fails**: requires Java 21 because Keycloak Config CLI depends on it.
- **SSL errors from external tools**: replace the generated certificate with one trusted by your OS or disable `-k` only after adding trust.
- **Credential request rejected**: confirm the client has the correct client scopes and attributes, and that the user possesses mapped claims (see protocol mappers inside each scope).

## More Docs

> - `src/` scripts - small, readable shell scripts for setup, configuration, and credential testing if you want to customize behavior.

Reviewer: Is this still necessary? If yes then maybe we should update the section title.

Author: Removed.

> That's it - a focused kit to light up OID4VC on Keycloak without extra ceremony.

Reviewer 2: ```suggestion
That's it - a focused kit to light up OID4VCI on Keycloak without extra ceremony.
```

Author: Fixed.

# Keycloak as an OID4VC Issuer

Bring OpenID for Verifiable Credential Issuance (OID4VCI) to a vanilla Keycloak with a single CLI. This repo contains a tiny-but-complete toolkit: Docker compose files, scripted setup, and curl-based request flows so you can evaluate VC issuance locally.

## What You Get
- Opinionated Keycloak instance (Postgres + HTTPS) with the `oid4vc-vci` feature enabled.
- `keycloak-ssi` CLI that wraps every script under `src/` and handles configuration, setup, and testing.

> - Ready-made credential definitions (IdentityCredential, SteuerberaterCredential, KMACredential) and demo users for both pre-authorized and authorization-code flows.

Reviewer: ```suggestion
- Ready-made credential definitions (IdentityCredential, SteuerberaterCredential, KMACredential) and a demo user for both pre-authorized and authorization-code flows.
```

Author: Fixed.

## TL;DR

> Prereqs: Java 17+ (Java 21 for `import`), `openssl`, `keytool`, `jq`, `yq`, `docker`, and optionally `figlet` for pretty banners.

### Quick Local Spin-Up (Docker Compose)
```bash
keycloak-ssi compose up -d   # boots Postgres + Keycloak with HTTPS on localhost
keycloak-ssi config          # configures realm, clients, demo users, keys

> keycloak-ssi test preauth IdentityCredential

Reviewer: As a new user, this command will fail because the CLI is not installed yet. So either use `./keycloak-ssi.sh ...` or mention that users need to ensure the CLI is installed before running this command.
Also, the following [comment](https://github.com/adorsys/keycloak-ssi-deployment/pull/189/files#r2556148224)
 may influence the changes here, so please consider it first.

**Just a suggestion:** instead of listing these commands here, why don't we redirect users to our CLI guide and ensure that the guide is clear enough for getting started? WDYT?

Author: I added a comment.

```
- Uses `config.yaml` plus overrides to drive runtime settings.

> - Logs land in `target/keycloak.log`; admin console available at `${KEYCLOAK_ADMIN_ADDR}` (defaults to `https://localhost:8443`).

Reviewer: I'm not sure the `compose` command actually writes logs to `target/keycloak.log`. Could you please verify this and update the documentation accordingly?

Author: Yes you are right, logs are not there 😬 corrected.

keycloak-ssi help
```
Installs into `~/.local/bin` and symlinks project files under `~/.local/share/keycloak-ssi-deployment`.

## How It Works

```
+-------------------+     +------------------------+
| keycloak-ssi CLI  |---->| src/deployment/*.sh    |--> Keycloak (oid4vc)
| - setup/config    |     | src/credentials/*.sh   |
| - compose bridge  |     |                        |
+-------------------+     +------------------------+
         ^                        ^
         | config.yaml            | overrides/env vars
         +------------------------+
```

1. `setup` builds Keycloak from either an official tarball or a source branch (controlled by `keycloak.version` and `keycloak.target_branch`).
2. `compose` runs the bundled `docker-compose.yml` (Postgres + Keycloak) when you prefer containers.
3. `config` applies the realm via Keycloak Config CLI (or kcadm) scripts, creating credential definitions as client scopes.
4. `test` scripts issue credentials through curl, covering pre-authorized and authorization-code + PKCE flows.

## Configuration Model

| Layer | Description |
| --- | --- |
| `config.yaml` | Base defaults for Keycloak, database, credentials, demo users. |

> | `config.override.yaml` | Optional environment-specific overrides checked into `.gitignore`. |

Reviewer: This file is missing in the `.gitignore`

Author: I removed the comment and will leave it out of the gitignore for reference. It doesn't contain sensitive data anyways.

| Environment vars | Highest priority. Use `SECTION_KEY=value` (e.g., `KEYCLOAK_HTTPS_PORT=9443`). Values inside YAML may reference `${ENV_VAR}`. |

Key settings to review:
- `keycloak.version` / `keycloak.target_branch`: decide between upstream tarball and custom branch (set version to `999.0.0-SNAPSHOT` to trigger branch build).
- `keycloak.https_port`, `keycloak.hostname`: align with your local HTTPS requirements.
- `database.*`: container DB credentials/ports if you reuse Postgres elsewhere.

> - `users.*`: passwords for demo users (never ship defaults to production).

Reviewer: The description here could be improved. Could you please refine it for clarity?

Author: I rewrote.

## Operating the CLI

| Command | What it does |
| --- | --- |
| `setup [-d]` | Builds / starts Keycloak with TLS using scripts under `src/deployment/0.*`. Pass `-d` to detach logging to `target/keycloak.log`. |
| `compose <args>` | Convenience wrapper for `docker compose` using the provided file and config expansion. |
| `config` | Runs `1.oid4vci_test_deployment.sh` and `2.configure_user_4_account_client.sh` to create realm, credential client scopes, and demo identities. Requires Keycloak to be reachable at `${KEYCLOAK_ADMIN_ADDR}`. |
| `import` | Uses Keycloak Config CLI to import a ready-made realm (needs Java 21). |
| `test <preauth|authcode> <CredentialType>` | Invokes curl-based flows inside `src/credentials/`. |
| `stop` | Stops a tarball-based server started via `setup`. |
| `install / uninstall` | Manage global availability of the CLI. |

> Use `keycloak-ssi help` for the built-in usage text (also mirrored in `docs/CLI_GUIDE.md`).

Reviewer: I don't think we should repeat this here.

Author: Okay, removed.

## Credential Request Playbooks

### Pre-Authorized Code + Key Binding
```bash
keycloak-ssi test preauth IdentityCredential
```
What happens:
1. Generates a wallet key pair.
2. Signs a proof of possession.
3. Calls the credential endpoint with the pre-authorized code provided during configuration.

### Authorization Code + PKCE
```bash
keycloak-ssi test authcode IdentityCredential
```
What happens:
1. Creates PKCE verifier/challenge.
2. Authenticates the demo user via the Keycloak auth endpoint.
3. Exchanges the auth code for a token using PKCE.

> 4. Requests IdentityCredential, SteuerberaterCredential, and KMACredential with key binding.

Reviewer: The behavior has changed: now, only the requested credential is returned. Using `keycloak-ssi test authcode IdentityCredential` will return just the `IdentityCredential`.

Author: Changed.

Both flows trust the self-signed TLS certificate generated by the setup scripts, so the curl helpers add `-k`. Swap in trusted certs for broader testing.

## Operational Notes
- **Verify Keycloak health**: `curl -k -s ${KEYCLOAK_ADMIN_ADDR}/realms/master` inside `cmd_setup` waits up to two minutes.
- **Certificates**: `generate-kc-certs.sh` produces the HTTPS cert/key and truststore referenced in `config.yaml`. Update passwords/paths there if you reuse existing certs.
- **Client scopes = credentials**: Each credential type is modeled as a client scope (see `src/config/client-scope-config.json`). Assign the scope to clients and add `"oid4vci.enabled": "true"` in client attributes.

> - **Logs & artifacts**: `target/` holds Keycloak logs, downloaded tarballs, and generated keys. Clean it if you need a fresh start.

Reviewer: ```suggestion
- **Logs & artifacts**: `target/` holds Keycloak logs, downloaded tarballs, and generated keys. **Clean it if you need a fresh start.**
```

Author: Improved.

## Troubleshooting
- **CLI cannot find the project**: run commands from the repo root or reinstall via `./keycloak-ssi.sh install`.

> - **Keycloak never becomes ready**: check `target/keycloak.log`, ensure ports (HTTPS / DB) are free, and confirm `openssl`, `keytool`, and Java are installed.

Reviewer: This applies only when starting Keycloak in the background using `./keycloak-ssi.sh setup -d`

Author: Removed since the logs are not there.

- **`keycloak-ssi import` fails**: requires Java 21 because Keycloak Config CLI depends on it.
- **SSL errors from external tools**: replace the generated certificate with one trusted by your OS or disable `-k` only after adding trust.
- **Credential request rejected**: confirm the client has the correct client scopes and attributes, and that the user possesses mapped claims (see protocol mappers inside each scope).

## More Docs

> - `docs/CLI_GUIDE.md` - extended CLI reference and command list.

Reviewer: I think it would make more sense to place this at the beginning of the document, as it’s the core guide that explains how to use the CLI. I believe some users may not read all the way to the end to find it.

Author: Okay, moved it above.
